### PR TITLE
initial concept

### PR DESCRIPTION
**Pull request by bradleybeddoes**
Currently looking for feedback on approach here. Thoughts on any test cases required would be useful too.

Have tested in real world scenario and all seems to work well.

Once we've gone around on this I will write the README before looking to merge this PR.

### CODE REVIEW

**Comment by bradleybeddoes** [comment]
Updated to work with Gotham [master](https://github.com/gotham-rs/gotham/commit/e62e1480f7e1839000cc3aac1cb1ac282102a2f2).

**Comment by bradleybeddoes** [comment]
n.b. temporary requirement for Gotham dependency to be located in source form at `../gotham`.

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    /// * An optional config instance. Utilises `r2d2::Config::default()` if not provided.
+    ///
+    /// # Panics
+    /// If the database identified in `database_url` cannot be connected to at application start.
+    ///
+    /// n.b. connection will be re-established if the database goes away and returns mid execution
+    /// without panic.
+    pub fn new(database_url: &str, c: Option<r2d2::Config<T, r2d2_diesel::Error>>) -> Self {
+        let manager = ConnectionManager::<T>::new(database_url);
+        let r2d2_config = match c {
```
Comment (ChristophWurst) [inline src/lib.rs]: FYI `c.unwrap_or(r2d2::Config::default())` would be shorter

https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or

Comment (bradleybeddoes) [reply src/lib.rs]: Nice find, a habit I need to get into for these situations, will address.

**Pull request merged**